File: runinator-command-center/src/core/domain/models/function/manifest.ts
// the `runinator-function.json` a package declares itself with, and the publish request it
// describes.
//
// mirrors `runinator-pack::functions::manifest`. the rules here are the ones the backend would
// reject on anyway — they are repeated so the dialog can say what is wrong before it uploads
// several megabytes of archive, not because the client is trusted to enforce them.

import type { FunctionExport, FunctionResourceLimits, FunctionRuntimeSpec } from "./index";

/// the alias a publish moves when the manifest names none.
export const DEFAULT_ALIAS = "latest";

export const MANIFEST_FILE = "runinator-function.json";

/// one export as a manifest declares it, before the server assigns it ids.
export type NewFunctionExport = Pick<FunctionExport, "name" | "handler"> &
  Partial<Pick<FunctionExport, "description" | "input" | "output">> & {
    limits?: FunctionResourceLimits;
  };

export interface FunctionManifest {
  name: string;
  namespace?: string | null;
  description?: string | null;
  runtime: FunctionRuntimeSpec;
  exports: NewFunctionExport[];
  /// the alias to move onto the published version. `null` publishes without moving anything, which
  /// is how a release is staged before promotion.
  alias?: string | null;
  /// extra paths left out of the archive. read by the compiler, not by a publish from a zip that is
  /// already built, so it is carried through untouched.
  exclude?: string[];
}

export interface NewFunctionPackage {
  name: string;
  namespace?: string | null;
  description?: string | null;
}

export interface NewFunctionVersion {
  package: NewFunctionPackage;
  artifact_digest: string;
  /// the manifest kept verbatim, so a republish can be compared against what was published.
  manifest: FunctionManifest;
  runtime: FunctionRuntimeSpec;
  exports: NewFunctionExport[];
  alias?: string | null;
}

/// the publish request a manifest describes, for an archive with the given digest.
export function publishRequest(
  manifest: FunctionManifest,
  artifactDigest: string,
): NewFunctionVersion {
  return {
    // the server stamps the caller's org; a manifest never names one, so a publish cannot land in
    // an org the publisher does not belong to.
    package: {
      name: manifest.name,
      namespace: manifest.namespace ?? null,
      description: manifest.description ?? null,
    },
    artifact_digest: artifactDigest,
    manifest,
    runtime: manifest.runtime,
    exports: manifest.exports,
    alias: manifest.alias === undefined ? DEFAULT_ALIAS : manifest.alias,
  };
}

/// read a manifest out of json text, rejecting anything that would publish something unaddressable
/// or uncallable.
export function parseManifest(text: string): FunctionManifest {
  let parsed: unknown;

  try {
    parsed = JSON.parse(text);
  } catch (error) {
    throw new Error(`${MANIFEST_FILE} is not valid json`, { cause: error });
  }

  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error(`${MANIFEST_FILE} must be a json object`);
  }

  const manifest = parsed as FunctionManifest;
  validateManifest(manifest);
  return manifest;
}

export function validateManifest(manifest: FunctionManifest): void {
  // the value arrives as parsed json, so nothing about its shape is guaranteed yet; it is read
  // through a partial view until each field has been checked.
  const declared = manifest as Partial<FunctionManifest>;
  validateIdent("package name", declared.name);

  if (declared.namespace) {
    validateIdent("namespace", declared.namespace);
  }

  const runtime: string | undefined = declared.runtime?.runtime;

  if (!runtime?.trim()) {
    throw new Error("runtime must name a runtime");
  }

  if (!Array.isArray(declared.exports) || declared.exports.length === 0) {
    throw new Error("a package must declare at least one export");
  }

  const seen = new Set<string>();

  for (const entry of declared.exports as Partial<NewFunctionExport>[]) {
    validateIdent("export name", entry.name);

    if (!entry.handler?.trim()) {
      throw new Error(`export '${entry.name}' has no handler`);
    }

    if (seen.has(entry.name)) {
      throw new Error(`export '${entry.name}' is declared twice`);
    }

    seen.add(entry.name);
  }

  if (declared.alias) {
    validateIdent("alias", declared.alias);
  }
}

// package, namespace, export, and alias names all become part of a dotted call path
// (`functions.<namespace>.<package>.<export>`), so they have to survive being split on `.`.
function validateIdent(what: string, value: string | undefined): asserts value is string {
  if (!value) {
    throw new Error(`${what} must not be empty`);
  }

  if (value.length > 64) {
    throw new Error(`${what} '${value}' is longer than 64 characters`);
  }

  if (!/^[A-Za-z]/.test(value)) {
    throw new Error(`${what} '${value}' must start with a letter`);
  }

  if (!/^[A-Za-z0-9_-]+$/.test(value)) {
    throw new Error(`${what} '${value}' may only contain letters, digits, '_', and '-'`);
  }
}

File: runinator-pack/src/functions/manifest.rs
//! the `runinator-function.json` manifest: what a package declares about itself.

use std::fs;
use std::path::Path;

use runinator_models::functions::{
    DEFAULT_ALIAS, FunctionRuntimeSpec, NewFunctionExport, NewFunctionPackage, NewFunctionVersion,
};
use runinator_models::value::Value;
use serde::{Deserialize, Serialize};

use crate::errors::{PackError, Result};

/// the manifest file a package directory is identified by.
///
/// json rather than toml so the pack crate keeps its dependency set and so schemas deserialize
/// straight into [`runinator_models::providers::ParameterMetadata`] — the same shape every other
/// manifest in the repo uses (`.rexrapm` is json too).
pub const MANIFEST_FILE: &str = "runinator-function.json";

/// a package's declared identity, runtime, exports, and what to archive.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FunctionManifest {
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub namespace: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub runtime: FunctionRuntimeSpec,
    pub exports: Vec<NewFunctionExport>,
    /// the alias to move onto the published version. defaults to `latest`; an explicit `null`
    /// publishes without moving anything, which is how a release is staged before promotion.
    #[serde(default = "default_alias")]
    pub alias: Option<String>,
    /// extra path patterns to leave out of the archive, on top of [`super::archive::DEFAULT_EXCLUDES`].
    #[serde(default)]
    pub exclude: Vec<String>,
}

fn default_alias() -> Option<String> {
    Some(DEFAULT_ALIAS.to_string())
}

impl FunctionManifest {
    /// read and validate the manifest in a package directory.
    pub fn load(directory: &Path) -> Result<Self> {
        let path = directory.join(MANIFEST_FILE);
        if !path.is_file() {
            return Err(PackError::source(format!(
                "{} is not a function package: no {MANIFEST_FILE}",
                directory.display()
            )));
        }
        let text = fs::read_to_string(&path)?;
        let manifest: Self = serde_json::from_str(&text)?;
        manifest.validate()?;
        Ok(manifest)
    }

    /// reject a manifest that would publish something unaddressable or uncallable.
    pub fn validate(&self) -> Result<()> {
        validate_ident("package name", &self.name)?;
        if let Some(namespace) = &self.namespace {
            validate_ident("namespace", namespace)?;
        }
        if self.runtime.runtime.trim().is_empty() {
            return Err(PackError::source("runtime must name a runtime"));
        }
        if self.exports.is_empty() {
            return Err(PackError::source(
                "a package must declare at least one export",
            ));
        }
        let mut seen = Vec::with_capacity(self.exports.len());
        for export in &self.exports {
            validate_ident("export name", &export.name)?;
            if export.handler.trim().is_empty() {
                return Err(PackError::source(format!(
                    "export '{}' has no handler",
                    export.name
                )));
            }
            if seen.contains(&export.name) {
                return Err(PackError::source(format!(
                    "export '{}' is declared twice",
                    export.name
                )));
            }
            seen.push(export.name.clone());
        }
        if let Some(alias) = &self.alias {
            validate_ident("alias", alias)?;
        }
        Ok(())
    }

    /// the publish request this manifest describes, for an archive with the given digest.
    pub fn publish_request(&self, artifact_digest: &str) -> NewFunctionVersion {
        NewFunctionVersion {
            package: NewFunctionPackage {
                name: self.name.clone(),
                namespace: self.namespace.clone(),
                description: self.description.clone(),
                // the server stamps the caller's org; a manifest never names one, so a pack cannot
                // publish into an org it does not belong to.
                org_id: None,
            },
            artifact_digest: artifact_digest.to_string(),
            manifest: self.as_value(),
            runtime: self.runtime.clone(),
            exports: self.exports.clone(),
            alias: self.alias.clone(),
        }
    }

    /// the manifest as stored verbatim on the version.
    pub fn as_value(&self) -> Value {
        serde_json::to_value(self)
            .map(Value::from)
            .unwrap_or(Value::Null)
    }

    /// the exports, sorted by name — the order the catalog and the cli present them in.
    pub fn sorted_exports(&self) -> Vec<NewFunctionExport> {
        let mut exports = self.exports.clone();
        exports.sort_by(|left, right| left.name.cmp(&right.name));
        exports
    }
}

// package, namespace, export, and alias names all become part of a dotted call path
// (`functions.<namespace>.<package>.<export>`), so they have to survive being split on `.`.
fn validate_ident(what: &str, value: &str) -> Result<()> {
    if value.is_empty() {
        return Err(PackError::source(format!("{what} must not be empty")));
    }
    if value.len() > 64 {
        return Err(PackError::source(format!(
            "{what} '{value}' is longer than 64 characters"
        )));
    }
    let head_is_alpha = value
        .chars()
        .next()
        .is_some_and(|character| character.is_ascii_alphabetic());
    if !head_is_alpha {
        return Err(PackError::source(format!(
            "{what} '{value}' must start with a letter"
        )));
    }
    let body_is_clean = value
        .chars()
        .all(|character| character.is_ascii_alphanumeric() || character == '_' || character == '-');
    if !body_is_clean {
        return Err(PackError::source(format!(
            "{what} '{value}' may only contain letters, digits, '_', and '-'"
        )));
    }
    Ok(())
}
